### ISSUE TITLE
A noobish question about training...

### ISSUE THREAD

**Issue by asigalov61** [lucidrains/sinkhorn-transformer#27]
@lucidrains 

First of all, everything seems to be working now in Google Colab so thank you very much for fixing it.

I have a quick question about training if you do not mind...

I get very high loss results. Here is the example:

training:   1%|          | 509/100000 [1:55:22<387:23:57, 14.02s/it]training loss: 2.495532512664795

Is this normal and I simply need to train more? Or does it mean that there is a problem somewhere?

2.49 in 2 hours is way too much IMHO. I am pretty sure I am not doing it right so your advice would be really apppreciated.

Thank you.

P.S. I am running your wiki8 example in Google Colab Pro.

**Comment by lucidrains** [comment]
@asigalov61 ohh, but it shows you've only run 509 steps, even in 2 hours? that doesn't seem quite right

how big are your batch sizes?

**Comment by asigalov61** [comment]
Yes, I know 500 steps are nothing but the loss would not go down so I did not see the point to continue running it...

Here are the settings:

NUM_BATCHES = int(1e5)
BATCH_SIZE = 28
GRADIENT_ACCUMULATE_EVERY = 4
LEARNING_RATE = 1e-4
VALIDATE_EVERY  = 100
GENERATE_EVERY  = 500
GENERATE_LENGTH = 512
SEQ_LEN = 4096

I increased the batches so I was wondering if the LR or anything else needs to be adjusted.

It may also be my code which I am checking right now, but I do not think it is causing the loss to stand still IMHO...

**Comment by lucidrains** [comment]
ohhh, keep the batch size at 4, and the learning rate at `3e-4`, and just leave it on overnight! it'll descend

it's because you are reaching for a sequence length of 4096

**Comment by asigalov61** [comment]
@lucidrains Thank you so much for your advice! I appreciate it a lot and I will do exactly what you suggest. I also found a few bugs within my code so hopefully correcting these things will help it to work properly.

To be honest, I am a bit disappointed by the performance of these latest SOTA transformers... They are slow and complex while not offering many advantages over the older implementations IMHO.

In fact, for my purposes, GPT2 remains unbeatable so I am yet to find a new thing to switch to :(

Anyways, thanks again.

Alex

**Comment by lucidrains** [comment]
@asigalov61 yes, I mostly agree with you! the ones I would recommend are https://github.com/lucidrains/routing-transformer and Longformer, which huggingface already has in their repository. Microsoft also has a sparse attention cuda kernel that is quite good https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/ops/sparse_attention/sparse_self_attention.py . Finally, on the cutting edge is https://github.com/lucidrains/performer-pytorch , which has potential, but still needs to prove itself (imo)

**Comment by asigalov61** [comment]
@lucidrains Yes, thank you again for your advice! Much appreciated.

I will try the Performer shortly and the Routing transformer as well. I know that its the latest and the greatest so to speak so def. worth a try...Its just I specialize in Music AI so not all implementations/approaches work well with music.

From my research and experience, I think I need something with sparse attention (thanks for recommending deep speed) and/or something with multiple embeddings (whatever that may mean as I am not sure). But if you have any ideas/sugestions, I would love to hear them.

Thank you.

**Comment by asigalov61** [comment]
@lucidrains Ok, so I think I got it to work in some capacity, but the loss is still pretty high for some reason. I did what you have suggested (4 batches/3-e4) but it did not help much. Perhaps I need to train longer as you have also suggested. However, I did get very good results with the Routing Transformer so I know that it is not a problem on my end IMHO.

I trained Routing and Sinkhorn for the same amount of time and here is what I've got:

Sinkhorn:

training:   0%|          | 327/100000 [41:06<198:57:29,  7.19s/it]training loss: 1.4748221635818481

Routing:

training:   1%|          | 1216/100000 [43:04<57:31:25,  2.10s/it]training loss: 0.5062560439109802

Same dataset and everything obviously.

Also, the Routing transformer showed much faster performance with the same 4096 seq. Why is it so?

Thanks.

**Comment by asigalov61** [comment]
@lucidrains Re-training the Sinkhorn properly right now and will post results shortly.

**Comment by asigalov61** [comment]
@lucidrains Tried training again but results were about the same. Maybe I used way too many batches? But with 4 batches it would not go down at all...This was trained on music so maybe why...Not sure...

NUM_BATCHES = int(1e5)
BATCH_SIZE = 28
GRADIENT_ACCUMULATE_EVERY = 4
LEARNING_RATE = 3e-4
VALIDATE_EVERY  = 100
GENERATE_EVERY  = 500
GENERATE_LENGTH = 1024
SEQ_LEN = 4096

Trained for 3 hours to 1.32 loss

Generation 2048 @ 0.8

The samples were not good at all :(

Perhaps this particular implementation does not like music if you are getting good results with say NLP...

[Music-Sinkhorn_MIDI_Output_Samples.zip](https://github.com/lucidrains/sinkhorn-transformer/files/6176054/Music-Sinkhorn_MIDI_Output_Samples.zip)
